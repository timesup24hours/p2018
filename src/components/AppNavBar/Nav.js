import React, { useEffect, useState, useRef } from 'react';
import { NavLink as Link } from 'react-router-dom';

import styles from './index.module.css';
import './nav.scss';

export default React.forwardRef((props, ref) => {
  const lineRef = useRef(null);
  const [style, setStyle] = useState({ left: 0, width: 0 });
  const getRect = (ele: Element | null) => {
    if (!ele) return 0;
    const rect = ele.getBoundingClientRect();
    return rect;
  };
  useEffect(() => {
    const activeNode: Element | null = document.querySelector(
      `.${styles.active}`
    );
    const left = getRect(activeNode).left;
    document.documentElement.style.setProperty('--nav-line-top', `${left}px`);
  });
  const handleNavOnClick = (event: React.MouseEvent) => {
    const ele: Element = event.target;
    if (!ele.classList.contains(styles.a)) return;
    const rect = getRect(ele.children[0]);
    setStyle(rect);
    // console.log('line: ', (lineRef.current.style.left = left));
    // document.documentElement.style.setProperty('--nav-line-top', `${left}px`);
  };
  return (
    <nav className={styles.nav} ref={ref}>
      <ul className={styles.nav_ul}>
        <li className={styles.nav_ul_li_1}>
          <Link
            exact
            className={styles.a}
            activeClassName={styles.active}
            onClick={event => {
              props.CustomhandleOnClick(event);
              handleNavOnClick(event);
            }}
            to="/"
          >
            <span>HOME</span>
          </Link>
        </li>
        <li className={styles.nav_ul_li_2}>
          <Link
            className={styles.a}
            activeClassName={styles.active}
            to="/my_notes"
            onClick={event => {
              props.CustomhandleOnClick(event);
              handleNavOnClick(event);
            }}
          >
            <span>NOTES</span>
          </Link>
        </li>
        <li className={styles.nav_ul_li_3}>
          <Link
            className={styles.a}
            activeClassName={styles.active}
            to="/css_effects"
            onClick={event => {
              props.CustomhandleOnClick(event);
              handleNavOnClick(event);
            }}
          >
            <span>UI</span>
          </Link>
        </li>
        <li className={styles.nav_ul_li_4}>
          <Link
            className={styles.a}
            activeClassName={styles.active}
            to="/contact_info"
            onClick={event => {
              props.CustomhandleOnClick(event);
              handleNavOnClick(event);
            }}
          >
            <span>CONTACT</span>
          </Link>
        </li>
        <div
          ref={lineRef}
          className={styles.line}
          style={{ left: style.left, width: style.width }}
        />
      </ul>
    </nav>
  );
});
