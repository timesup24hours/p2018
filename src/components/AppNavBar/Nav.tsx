import React, { useEffect, useState, useRef, MouseEvent } from 'react';
import { NavLink as Link } from 'react-router-dom';

// import styles from './index.module.css';
import './nav.scss';

interface rectObject {
  left: number;
  top: number;
  width: number;
  bottom: number;
}

interface NavProps {
  customhandleOnClick: (event: MouseEvent) => void;
}

export default React.forwardRef(
  (props: NavProps, ref?: React.Ref<HTMLDivElement>) => {
    const lineRef = useRef<HTMLDivElement>(null);
    const [style, setStyle] = useState({
      left: 0,
      width: 0,
      bottom: 36,
      top: 0
    });
    const getRect = (ele: Element | null): ClientRect | rectObject => {
      if (!ele) return { left: 0, width: 0, bottom: 36, top: 0 };
      const rect = ele.getBoundingClientRect();
      return rect;
    };
    useEffect(() => {
      const activeNode: Element | null = document.querySelector(`.active`);
      if (!activeNode) return;
      const rect = getRect(activeNode!.children[0]);
      document.documentElement.style.setProperty(
        '--nav-line-left',
        `${rect.left}px`
      );
      document.documentElement.style.setProperty(
        '--nav-line-width',
        `${rect.width}px`
      );
      document.documentElement.style.setProperty(
        '--nav-line-bottom',
        `${rect.bottom - 1}px`
      );
    });
    // const handleNavOnClick = (event: React.MouseEvent) => {
    //   const ele = event.currentTarget!;
    //   if (!ele.classList.contains('a')) return;
    //   const rect = getRect(ele.children[0]);
    //   setStyle(rect);
    //   document.documentElement.style.setProperty(
    //     '--nav-line-top',
    //     `${rect.left}px`
    //   );
    // };
    return (
      <nav className="Nav_nav" ref={ref}>
        <ul className="nav_ul">
          <li className="nav_ul_li_1">
            <Link
              exact
              className="a"
              activeClassName="active"
              onClick={event => {
                props.customhandleOnClick(event);
                // handleNavOnClick(event);
              }}
              to="/"
            >
              <span>HOME</span>
            </Link>
          </li>
          <li className="nav_ul_li_2">
            <Link
              className="a"
              activeClassName="active"
              to="/my_notes"
              onClick={event => {
                props.customhandleOnClick(event);
                // handleNavOnClick(event);
              }}
            >
              <span>NOTES</span>
            </Link>
          </li>
          <li className="nav_ul_li_3">
            <Link
              className="a"
              activeClassName="active"
              to="/css_effects"
              onClick={event => {
                props.customhandleOnClick(event);
                // handleNavOnClick(event);
              }}
            >
              <span>UI</span>
            </Link>
          </li>
          <li className="nav_ul_li_4">
            <Link
              className="a"
              activeClassName="active"
              to="/contact_info"
              onClick={event => {
                props.customhandleOnClick(event);
                // handleNavOnClick(event);
              }}
            >
              <span>CONTACT</span>
            </Link>
          </li>
          <div
            ref={lineRef}
            className="nav_line"
            // style={{ left: style.left, width: style.width }}
          />
        </ul>
      </nav>
    );
  }
);
