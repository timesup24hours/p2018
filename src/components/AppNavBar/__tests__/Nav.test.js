import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';

import Nav from '../Nav.tsx';

describe('Nav', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<Nav debug />);

    expect(component).toMatchSnapshot();
  });

  it('should be able to call the fn when click', () => {
    const customhandleOnClick = jest.fn();
    const component = shallow(
      <Nav customhandleOnClick={customhandleOnClick} />
    );
    // console.log(component.find('[data-testid="my_notes"]').debug());
    const link = component.find('[data-testid="my_notes"]');
    link.simulate('click');

    expect(customhandleOnClick).toHaveBeenCalled();

    // expect(component.find('.active').length).toBe(1);
    // console.log(component.find('.active').debug());
    // console.log(link.debug());
    // expect(component).toMatchSnapshot();
    // component.unmount();

    // expect(link.hasClass('active')).toBe(true);
  });

  it('should be able to call the fn when click', () => {
    const customhandleOnClick = jest.fn();
    const component = mount(
      <Router>
        <Nav customhandleOnClick={customhandleOnClick} />
      </Router>
    );
    const link = component.find('[data-testid="my_notes"]').at(0);
    link.simulate('click');

    // console.log(component.find('.active').debug());
    // expect(link.hasClass('active')).toBeTruthy();
    expect(link.text()).toBe('NOTES');
    expect(component.find('.active').length).toBe(2);
  });
});
