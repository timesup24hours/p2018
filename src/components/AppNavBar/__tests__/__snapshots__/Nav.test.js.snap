// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should render correctly in "debug" mode 1`] = `
<nav
  className="Nav_nav"
>
  <ul
    className="nav_ul"
  >
    <li
      className="nav_ul_li_1"
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="a"
        exact={true}
        onClick={[Function]}
        to="/"
      >
        <span>
          HOME
        </span>
      </NavLink>
    </li>
    <li
      className="nav_ul_li_2"
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="a"
        data-testid="my_notes"
        onClick={[Function]}
        to="/my_notes"
      >
        <span>
          NOTES
        </span>
      </NavLink>
    </li>
    <li
      className="nav_ul_li_3"
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="a"
        onClick={[Function]}
        to="/css_effects"
      >
        <span>
          UI
        </span>
      </NavLink>
    </li>
    <li
      className="nav_ul_li_4"
    >
      <NavLink
        activeClassName="active"
        aria-current="page"
        className="a"
        onClick={[Function]}
        to="/contact_info"
      >
        <span>
          CONTACT
        </span>
      </NavLink>
    </li>
    <div
      className="nav_line"
    />
  </ul>
</nav>
`;
